File: javascript/test/callbackParser.test.js
import {loadTestFile} from '../test/testUtils.js';
import {strict as assert} from 'assert';
import {FountainCallbackParser} from "../src/callbackParser.js";

describe('Simple Callback Parser', () => {
  it('should work', () => {

    const match = loadTestFile("SimpleCallbackParser.txt");

    let out = [];

    let fp = new FountainCallbackParser();
    fp.onDialogue = (character, extension, parenthetical, line, isDualDialogue) => {
        out.push("DIALOGUE:"+
            ` character:${character}`+
            ` extension:${extension}`+
            ` parenthetical:${parenthetical}`+
            ` line:${line}`+
            ` dual:${isDualDialogue}`);
    }
    fp.onAction = (text) => {
        out.push(`ACTION: text:${text}`);
    }

    fp.onSceneHeading = (text, sceneNumber) => {
        out.push(`HEADING: text:${text} sceneNum:${sceneNumber}`);
    }

    fp.onLyrics = (text) => {
        out.push(`LYRICS: text:${text}`);
    }

    fp.onTransition = (text) => {
        out.push(`TRANSITION: text:${text}`);
    }

    fp.onSection = (text, level) => {
        out.push("SECTION:"+
            ` level:${level}`+
            ` text:${text}`);
    }

    fp.onSynopsis = (text) => {
        out.push(`SYNOPSIS: text:${text}`);
    }

    fp.onPageBreak = () => {
        out.push("PAGEBREAK");
    }

    fp.onTitlePage = (entries) => {
        let page = "TITLEPAGE:";
        for (const entry of entries) {
            page+=` ${entry.key}:${entry.value}`;
        }
        out.push(page);
    }

    fp.ignoreBlanks = true;

    fp.addText(loadTestFile("TitlePage.fountain"));
    fp.addText(loadTestFile("Sections.fountain"));
    fp.addText(loadTestFile("Character.fountain"));
    fp.addText(loadTestFile("Dialogue.fountain"));

    let output = out.join("\n");
    //console.log(output);
    assert.equal(output, match);
  });
});

File: csharp/ScreenplayTools.Tests/Fountain/CallbackParserTest.cs
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

namespace ScreenplayTools.Tests;
using System.IO;
using ScreenplayTools;
using ScreenplayTools.Fountain;

public class CallbackParserTest
{
    private string loadTestFile(string fileName) {
        return File.ReadAllText("../../../../../tests/"+fileName);
    }

    private string asNull(string? value) {
        if (value==null)
            return "null";
        return value;
    }

    [Fact]
    public void TestParse()
    {
        string match = loadTestFile("SimpleCallbackParser.txt");

        var outLines = new List<string>();

        CallbackParser fp = new CallbackParser();

        fp.OnDialogue = (Character, Extension, Parenthetical, Line, IsDualDialogue) => {
            outLines.Add("DIALOGUE:"+
                " character:"+Character+
                " extension:"+asNull(Extension)+
                " parenthetical:"+asNull(Parenthetical)+
                " line:"+Line+
                " dual:"+IsDualDialogue);
        };

        fp.OnAction = (Text) => {
            outLines.Add("ACTION: text:"+Text);
        };

        fp.OnSceneHeading = (Text, SceneNumber) => {
            outLines.Add("HEADING: text:"+Text+" sceneNum:"+asNull(SceneNumber));
        };

        fp.OnLyrics = (Text) => {
            outLines.Add("LYRICS: text:"+Text);
        };

        fp.OnTransition = (Text) => {
            outLines.Add("TRANSITION: text:"+Text);
        };

        fp.OnSection = (Text, Level) => {
            outLines.Add("SECTION: level:"+Level+" text:"+Text);
        };

        fp.OnSynopsis = (Text) => {
            outLines.Add("SYNOPSIS: text:"+Text);
        };

        fp.OnPageBreak = () => {
            outLines.Add("PAGEBREAK");
        };

        fp.OnTitlePage = entries => {
            var page = "TITLEPAGE:";
            foreach (var entry in entries)
            {
                page+=$" {entry.Key}:{entry.Value}";
            }
            outLines.Add(page);
        };

        fp.IgnoreBlanks = true;

        fp.AddText(loadTestFile("TitlePage.fountain"));
        fp.AddText(loadTestFile("Sections.fountain"));
        fp.AddText(loadTestFile("Character.fountain"));
        fp.AddText(loadTestFile("Dialogue.fountain"));

        string output = string.Join("\n", outLines);
        output = output.Replace("False", "false");
        output = output.Replace("True", "true");

        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }
}